src/array.h

```diff
@@ -5,31 +5,38 @@
 #include <array>
 
 namespace Rcpp {
-namespace traits {
-
-// Partial specialisation to allow as<array<T,D>>(...)
-template <typename ElementType, int Dimensionality>
-class Exporter<std::array<ElementType,Dimensionality>>
-{
-private:
-    std::array<ElementType,Dimensionality> value;
-
-public:
-    Exporter (SEXP x)
-    {
-        std::vector<ElementType> vec = as<std::vector<ElementType>>(x);
-        if (vec.size() != Dimensionality)
-            throw Rcpp::exception("Array does not have the expected number of elements");
-        for (int i=0; i<Dimensionality; i++)
-            value[i] = vec[i];
-    }
-    
-    std::array<ElementType,Dimensionality> get () { return value; }
-};
-
-} // traits namespace
-} // Rcpp namespace
+  namespace traits {
+
+    // Partial specialisation to allow as<array<T,D>>(...)
+    template <typename T, std::size_t D> class Exporter<std::array<T,D>>;
+
+  }
+}
 
 #include <Rcpp.h>
 
+namespace Rcpp {
+  namespace traits {
+
+    template <typename T, std::size_t D> class Exporter<std::array<T,D>> {
+      typedef typename std::array<T,D> ATD;
+
+      // Convert the type to a valid rtype.
+      const static int RTYPE = Rcpp::traits::r_sexptype_traits<T>::rtype;
+      Rcpp::Vector<RTYPE> vec;
+
+    public:
+      Exporter(SEXP x): vec(x) {
+        if (TYPEOF(x) != RTYPE) Rcpp::stop("Wrong R type for array");
+      }
+      ATD get() {
+        ATD x;                  // resize step neeed? or implicit because of D?
```

Comment:
No resize step is necessary for `array`, no

Reply:
I have come to the same conclusion -- because 'D' (or maybe called 'N") is there.

This was a useful exercise. I probably add `wrap()` as well and make this available somewhere.